### ISSUE TITLE
Alerts Tree Update takes a LONG time after spider completes if LOTS of issues found

### ISSUE THREAD

**Issue by davewichers** [zaproxy/zaproxy#3881]
I'm running a spider against a site with apparently LOTS of pages, Over 27,000 according to ZAP. 99+% of these pages are static HTML documentation pages, which I can of course exclude from a spider if I wanted to.

However, it's not the spidering that takes a long time, I think it's the updating of the Alerts tree. The spider tab says 19000+ Nodes added. Is that nodes to the Site tree? Nodes to the Alerts tree? Not clear. I'm guessing Site tree since there are currently 40,000 issues and still growing reported in the Alerts tree (across 21 different categories).

I'm watching 5 different Alert node counters count up one at a time (e.g., X-Frame-Options header missing currently at 5400+ and still growing).

Note that I have all the alpha passive rules turned on (not the beta), because I was testing out the alpha Cross-Domain configuration rule. As such I might have far more rules than normal turned on.

I'm wondering if the Alerts tree update can be made more efficient when large numbers of alerts are generated like this? For example, can we add large batches of nodes to the alerts tree and then only update the UI once in a while? (I.e., it appears that EVERY TIME a new alert is added to the alerts tree, the entire tree is updated in the UI so one node will increment its count by 1). I'm just speculating that this is the cause of why its taking so darn long.

Any other ideas most welcome.

And by the time I finished writing this, my alert count is now up to 8700 on some nodes, and still going up :-(

**Comment by psiinon** [comment]
I can believe this happens.
At one stage I planned to allow users to configure a maximum number of alerts that would be checked for. We already support this in the reports, but if we also supported this in the scanning then we could skip rules that have exceeded that maximum. This could speed up ZAP scanning as a side effect.
However its non trivial to work out the interface for this - is it across all sites, just those in scope...
I also agree that we should make the Sites tree (and all of ZAP) more efficient and effective for scanning large sites - the above could be part of it, but definitely not the whole story.

**Comment by davewichers** [comment]
Does the internal API for updating the Alerts tree only allow 1 alert to be added at a time? And every time it is called, the UI updates automatically? Or are there options for either:
a) Adding multiple nodes at once (then updating the UI once), or
b) Adding a node, but don't update the UI. Adding another node the same way, then manually at the end, or periodically, thru another call, causing the UI to update?

By the way, At the end I had like 26K issues reported in 5 different categories (all missing / wrong header rules) out of about 27K pages found.

**Comment by davewichers** [comment]
Its been 7 months since this issue was reported. I'd like this issue to just focus on improving the efficiency of updating the Alerts tree when LOTS of issues are found in a short period of time. And not make it more complex by adding configuration options to anything.  I have to believe it can't be that hard to figure out a way to update the Alerts tree more efficiently in this situation. It's easy to repeat this behavior, by the way. Simply spider OWASP Benchmark starting at the root of the application, and I think you'll end up with over 7000 issues from the default rules in 4 different categories. Any chance someone can work on this? Thx.

**Comment by kingthorin** [comment]
Hi Dave, as you know this is an open source project. We have issues dating back 6 years. We'd love for them to all be addressed in short order but with a limited pool of regular contributors that just isn't possible. 

We're happy to see new contributors any time. If this is something you or someone at your company would be interested in tackling let us know.

**Comment by davewichers** [comment]
I understand and appreciate the entire ZAP team’s hard work. I’m just trying to give this particular issue a ‘nudge’.

**Comment by psiinon** [comment]
Always worth a try ;)

**Comment by kingthorin** [comment]
How long does it take to spider the 27,000 URLs?

I suspect what you’re encountering is not necessarily an issue of UI updates but rather the passive scanning of 27k+ requests/responses.

**Comment by davewichers** [comment]
No. The scan completes much faster than the UI update. After the scan is
complete, the updates on 6+ nodes that all have to increment one at a time
up to a count of 27000+ is what takes lots of time. You can see them just
spinning upwards for 10-15+ minutes after the scan is complete.

**Comment by davewichers** [comment]
To me, this issue is entirely in the efficiency of the sites tree update,
per Simon's original comment: "I also agree that we should make the Sites
tree (and all of ZAP) more efficient and effective for scanning large sites
- the above could be part of it."

Seems like there should be a way to update the model of the sites tree, but
not the view, except  maybe once every 100 updates, or so. I would think
that would be make the update of the sites view MUCH faster when there are
10s of thousands of values to add to the tree. In my example, it approaches
200,000 total.

**Comment by kingthorin** [comment]
Completion of the spider does not necessarily reflect completion of all the passive scanners.

**Comment by thc202** [comment]
Yes, the scanners are still scanning when that happens.

**Comment by psiinon** [comment]
How about we have a threshold on the passive scan queue - if it exceeds the (configurable) threshold then we only update the sites tree every 100 (again, configurable) alerts.
Or something like that :)

**Comment by davewichers** [comment]
That would be great, and I suspect it would help a lot.

**Comment by kingthorin** [comment]
I'll look into getting a animated gif of it all.

**Comment by thc202** [comment]
Even though it's not a tree update issue (currently) it's UI related (the passive scan thread ends up waiting for the EDT).

**Comment by kingthorin** [comment]
Here's an animated gif. Note I did pause the recording at various points so that it could still be illustrative without being crazy long to watch. Keep on eye on the time stamps and counts. (I dunno why I bothered to set max children at the start, because it never comes into play, but I don't have time to go through recording it all over again...)

I suppose there could be an impact on passive scanning from UI updates, however it "seems" to be on the order of ms or ns over minutes worth of passive scanning.

![3881_cap](https://user-images.githubusercontent.com/7570458/40733625-193626d8-6404-11e8-8d47-25d01bb0c827.gif)
